### PR TITLE
Discard first random value after setting seed

### LINKED ISSUE #3753: seeding math.randomseed() with similar values give similar random results (DEF-2847)
# Description:

Seeding the randomizer with similar values generates initial random start values that are close. This is especially troublesome when randomizing within ranges and seeding from the clock, which inreases with 1 each second.

# Steps to reproduce:

math.randomseed(0)
print(math.random()) --> 0.00115966796875
math.randomseed(1)
print(math.random()) --> 0.001251220703125
math.randomseed(2)
print(math.random()) --> 0.001373291015625
math.randomseed(3)
print(math.random()) --> 0.00146484375

and even worse:

math.randomseed(0)
print(math.random(1, 10)) --> 1
math.randomseed(10)
print(math.random(1, 10)) --> 1
math.randomseed(200)
print(math.random(1, 10)) --> 1
math.randomseed(3000)
print(math.random(1, 10)) --> 4  (finally!)

2017-08-21 11:35:39 (mikael.saker)
More info:

(LINK REMOVED)
2017-08-21 11:43:29 (mikael.saker)
Lua 5.2 fixes this by a call to rand() once in math_randomseed(). Propose we patch our Lua with this:

{code}
static int math_randomseed (lua_State *L) {
  srand(luaL_checkunsigned(L, 1));
  (void)rand(); /* discard first value to avoid undesirable correlations */
  return 0;
}
{code}

### LINKED ISSUE #6170: Make math.random developer-friendly
**Is your feature request related to a problem? Please describe (REQUIRED):**
Randomness and initialization of `math.random` is the most popular question in the Defold chat in Telegram. 
Every time newcomers get the answers like:
1. Call [`math.random` three times](http://lua-users.org/wiki/MathLibraryTutorial) after `math.randomseed` to get real random values (on Windows at least). It sounds like black magic 😄 
2. Use https://github.com/selimanac/defold-random or https://github.com/indiesoftby/defold-splitmix64
etc.

I do think that Defold's `math.random` should be developer-friendly out of the box. 

**Describe the solution you'd like (REQUIRED):**
Substitute Lua's `math.random` + `math.randomseed` with any simple PRNG (xoroshiro128+, [splitmix64](https://github.com/indiesoftby/defold-splitmix64/blob/main/splitmix64/src/main.cpp), there are lots!) to make it cross-platform and to not rely on weird OS implementations.

### PR DESCRIPTION
**Pull request by britzl**
Fixes #3753
Fixes #6170

### CODE REVIEW

File: engine/script/src/script.cpp [hunk 1]
```diff
@@ -139,6 +139,8 @@ namespace dmScript
         lua_getglobal(L, RANDOM_SEED);
         uint32_t* seed = (uint32_t*) lua_touserdata(L, -1);
         *seed = luaL_checkint(L, 1);
+        // discard first value to avoid repeated values
```
Comment (britzl) [inline engine/script/src/script.cpp:142]: This is what's done in Lua 5.2 for instance:

https://github.com/lua/lua/blob/v5-2/lmathlib.c#L228

Comment (JCash) [reply engine/script/src/script.cpp:142]: Is this needed after we set the seed in `math.randomseed(seed)`?

Comment (britzl) [reply engine/script/src/script.cpp:142]: I'm not sure I understand?

Comment (JCash) [reply engine/script/src/script.cpp:142]: I'm not sure either. I really thought this code was inside the "math.rand()" fuction 🤔 

But still, the question I intended is still valid.
We have two use cases:

1) call math.rand()  
2) call math.randomseed() and then math.rand()

This PR only pre-warms the 2) case, inside the math.randomseed().
So, if you're using 1), you'll need to do "math.rand()" yourself in your script?

I.e. do we need a dmMath::Rand(seed); inside the [dmScript::Initialize()](https://github.com/defold/defold/blob/dev/engine/script/src/script.cpp#L179) as well?

Comment (britzl) [reply engine/script/src/script.cpp:142]: Ah, ok, I understand. I think it is much less of a problem when you call math.random() without first setting a seed. But yeah, maybe it's best to also call dmMath::Rand(seed) once during init.

Comment (JCash) [reply engine/script/src/script.cpp:142]: Not sure how it's less of a problem?

The seed is just an external integer, which is then passed to the dmMath::Rand() function.
The Rand function doesn't care if the seed is 0 or not.

So, if seed!=0 is a problem, then I assume seed=0 is a problem too.

Comment (JCash) [reply engine/script/src/script.cpp:142]: I think my point is supported by this issue (i.e. the default seed=0 is low, thus will generate issues):
https://github.com/defold/defold/issues/3753

Also, is calling it +1 times fixing the observed behavior in #3753?

**Comment by JCash** [comment]
A bigger discussion, which I didn't see in the issues #3753 nor #6170, is about backwards compatibility.

It's easy to say "it's rand(), it doesn't matter", but it uses a developer chosen seed, and the developer might rely on the outcome of that function.

E.g.
```
level1_seed=123456 -- generates a nice level 1
level2_seed=456789 -- generates a nice level 2

generate_level(level1_seed)
```

Changing the outcome of the math.random() function, might a lot for the users.

Is this backwards compatibility change considered?

File: engine/script/src/script.cpp (hunk 1 above)
Comment (britzl) [reply engine/script/src/script.cpp:142]: Yes, calling it once is sufficient according to my own tests.

**Comment by britzl** [comment]
> Changing the outcome of the math.random() function, might change a lot for the users.
> 
> Is this backwards compatibility change considered?

I considered it and decided that the problem we solve is much greater than the possible problem we create.

Some alternatives:

* Do nothing and recommend developers to manually call math.random() once or twice after setting the seed
* Do nothing and suggest developers to use a different PRNG (perhaps an extension) that doesn't have this issue
* Add an option to math.randomseed(seed, number_of_values_to_discard) to discard a configurable number of values. Backwards compatible but leaving default implementation broken
* Add new math.randomseed2(seed) that does the right thing and discards one or more value

**Comment by JCash** [comment]
Ok. I wanted that design choice be stated explicitly so that we can refer to it later.

I think we can use the extra option to randomseed() or the "math.randomseedold()" approach if it turns out to be a problem.

File: engine/script/src/test/test_script.cpp [hunk 2]
```diff
@@ -215,9 +215,9 @@ TEST_F(ScriptTest, TestRandom)
 {
     int top = lua_gettop(L);
     ASSERT_TRUE(RunString(L, "math.randomseed(123)"));
-    ASSERT_TRUE(RunString(L, "assert(math.random(0,100) == 1)"));
     ASSERT_TRUE(RunString(L, "assert(math.random(0,100) == 58)"));
-    ASSERT_TRUE(RunString(L, "assert(math.abs(math.random() - 0.70419311523438) < 0.0000001)"));
```
Comment (JCash) [inline engine/script/src/test/test_script.cpp:220]: We don't want a test for avoiding the previous behavior?

(The current test only tests the actual values, not the series of values described in the issue #3753)

**Review by JCash — DISMISSED** [review]

File: engine/script/src/test/test_script.cpp [hunk 3]
```diff
@@ -215,9 +215,21 @@ TEST_F(ScriptTest, TestRandom)
...
+
+    // https://github.com/defold/defold/issues/3753
+    const char *sum_random_numbers =
+        "local sum = 0\n"
+        "local count = 1000\n"
+        "for i=1,count do\n"
+        "    math.randomseed(i)\n"
+        "    sum = sum + math.random(1, 10)\n"
+        "end\n"
+        "assert(math.floor(sum/count) == 5)\n";
```
Comment (britzl) [inline engine/script/src/test/test_script.cpp:230]: Before this change this test would have resulted in an average floored value of 1

**Review by JCash — APPROVED** [review]

**Pull request merged**